File: app/src/main/java/tiiehenry/android/app/snapshotor/config/ShotConfig.java
package tiiehenry.android.app.snapshotor.config;

import com.alibaba.fastjson2.JSON;
import com.alibaba.fastjson2.TypeReference;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;


/**
 * 快照配置类
 * 纯数据类，使用 fromJson/toJson 进行序列化/反序列化
 */
public class ShotConfig {
    // 配置字段，直接访问
    public String compressAlgorithm = "";
    public Set<String> compressItems = new HashSet<>(CompressItems.getAll());
    public boolean permission = true;
    public String extraItems = "";
    public boolean autoSnapshot = false;
    public boolean uninstallArchived = false;

    /**
     * 从 JSON 字符串解析配置（静态工厂方法）
     */
    public static ShotConfig fromJson(String jsonString) {
        return JSON.parseObject(jsonString, ShotConfig.class);
    }

    /**
     * 将配置转换为 JSON 字符串
     */
    public String toJson() {
        return JSON.toJSONString(this);
    }

    /**
     * 获取解析后的额外压缩项列表
     */
    public List<ExtraCompressItem> getExtraItemsList() {
        if (extraItems == null || extraItems.isEmpty()) {
            return new ArrayList<>();
        }
        try {
            return JSON.parseObject(extraItems, new TypeReference<List<ExtraCompressItem>>() {});
        } catch (Exception e) {
            return new ArrayList<>();
        }
    }

    /**
     * 设置额外压缩项列表
     */
    public void setExtraItemsList(List<ExtraCompressItem> items) {
        this.extraItems = JSON.toJSONString(items);
    }

    /**
     * 是否有压缩算法配置
     */
    public boolean hasCompressAlgorithm() {
        return compressAlgorithm != null && !compressAlgorithm.isEmpty();
    }

    /**
     * 是否有压缩项配置
     */
    public boolean hasCompressItems() {
        return compressItems != null && !compressItems.isEmpty();
    }
}

File: app/src/main/java/tiieherny/android/app/snapshotor/config/ShotConfig.kt
package tiieherny.android.app.snapshotor.config

import com.alibaba.fastjson2.JSON
import com.alibaba.fastjson2.TypeReference
import com.tencent.mmkv.MMKV

class ShotConfig(private val mmkv: MMKV) {
    private companion object {
        const val KEY_COMPRESS_ALGORITHM = "compressAlgorithm"
        const val KEY_COMPRESS_ITEMS = "compressItems"
        const val KEY_PERMISSION = "permission"
        const val KEY_EXTRA_ITEMS = "extraItems"
        const val KEY_AUTO_SNAPSHOT = "autoSnapshot"
        const val KEY_UNINSTALL_ARCHIVED = "uninstallArchived"
    }

    var autoSnapshot: Boolean
        get() = mmkv.decodeBool(KEY_AUTO_SNAPSHOT, false)
        set(value) {
            mmkv.encode(KEY_AUTO_SNAPSHOT, value)
        }

    var compressAlgorithm: String
        get() = mmkv.decodeString(KEY_COMPRESS_ALGORITHM) ?: ""
        set(value) {
            mmkv.encode(KEY_COMPRESS_ALGORITHM, value)
        }

    var compressItems: Set<String>
        get() = mmkv.decodeStringSet(KEY_COMPRESS_ITEMS) ?: CompressItems.all
        set(value) {
            mmkv.encode(KEY_COMPRESS_ITEMS, value)
        }

    var permission: Boolean
        get() = mmkv.decodeBool(KEY_PERMISSION, true)
        set(value) {
            mmkv.encode(KEY_PERMISSION, value)
        }

    var extraItems: List<ExtraCompressItem>
        get() {
            val json = mmkv.decodeString(KEY_EXTRA_ITEMS) ?: return emptyList()
            return try {
                JSON.parseObject(json, object : TypeReference<List<ExtraCompressItem>>() {})
            } catch (e: Exception) {
                emptyList()
            }
        }
        set(value) {
            mmkv.encode(KEY_EXTRA_ITEMS, JSON.toJSONString(value))
        }

    var uninstallArchived: Boolean
        get() = mmkv.decodeBool(KEY_UNINSTALL_ARCHIVED, false)
        set(value) {
            mmkv.encode(KEY_UNINSTALL_ARCHIVED, value)
        }
}